docker-compose.yml

```diff
@@ -11,10 +11,24 @@ services:
         volumes:
           - .:/go/src/github.com/letsencrypt/boulder
           - /tmp:/tmp
-        network_mode: bridge
+        network_mode: service:bhsm
+        depends_on:
+          - bhsm
+          - bmysql
+    bhsm:
```

Comment:
I think the rearrangement of the `docker-compose.yml` ended up putting the `ports` and `extra_hosts` for the `boulder` container onto the `bhsm` container.

`docker-compose ps` after a `docker-compose up` shows the `boulder_bhsm_1` container has all of the exposed ports, not the `boulder_boulder_1` container.

Am I misunderstanding an intentional change? I'm the first to admit I find Docker mysterious on the best of days :laughing: I was able to issue a certificate successfully from my host system against the Boulder container using Certbot with this unmodified `docker-compose.yml`, so maybe I am.

Reply:
I see the comment now: " # extra_hosts and ports for boulder are declared here as other containers uses the networking stack of this container." which explains the reason the ports show up there.

@ibukanov Can you expand on why the other containers use this container's network stack? Without knowing more it seems more intuitive to have the `boulder` container be the primary container with the network stack definitions but I'm likely overlooking the reason you did things this way.

Reply:
@cpu The reason for using the network stack of bhsm and not boulder is dependencies. The network stack requires that container runs before the containers that use it are started. So using the stack of the boulder container requires that the boulder container always starts with bhsm.

I also tried to use docker networking rather the stack reuse, but that failed the tests.

Reply:
Regarding that test failure without network stack sharing. If I just connect all docker containers to the docker network that compose creates, then ocsp-updater test fails. I suspect the reason it works with the legacy bridge networking is that --link puts the container ip addresses into /etc/hosts while to find container names with Docker networking one has to use Docker DNS proxy from /etc/resolv.conf. As tests uses own DNS, that does not work.

Reply:
@cpu When one shares the network stack between containers, one has to expose ports, define DNS aliases etc in the container which stack is shared.

Reply:
@ibukanov Thanks for the explanation - that makes sense :-)